### DISCUSSION TITLE
Where should complexity live?

### DISCUSSION THREAD

**Discussion by jinjagit** [oliverbarnes/liquid-voting-service#236] [Ideas]
Currently, our api has the most complex operations related to delegations in the _core_ layer. E.g. creating delegations involves upserting participants, voting_methods and delegations, and selects some, all or none of these depending on different contexts (global vs proposal specific, for example) all within the `lib/liquid_voting/delegations.ex` file - within what I call the 'core' level.

Conversely, we have the most complex operations related to votes in the _absinthe_ layer. E.g. creating votes involves upserting participants, etc. (many similar complexities to creating delegations, with some differences), mostly within the `lib/liquid_voting_web/resolvers/voting.ex` file.

This seems like a somewhat arbitrary choice, that has been developed in _both_ directions.

In the interests of making our codebase easier to understand and develop, I think we should decide on the best layer to locate the more complex logic. I don't, however, feel the resulting refactoring is urgent at this point.

Thoughts?

**Comment by oliverbarnes** [comment]
Good question :) in principle the business logic layer on Phoenix is the Context: https://hexdocs.pm/phoenix/contexts.html - so the absinthe/resolvers layer should forward to a Context to take care of the nitty gritty.

This issue from a few months ago is related to this, too: https://github.com/liquidvotingio/api/issues/155

**Comment by niklaslong** [comment]
I like this question! First off, I'd agree that the refactor isn't urgent. The code is fairly readable and there's only a slight blurring of concerns at this point imo. The simple "endpoints" are calling on the standard CRUD functions in the `Voting` module, no real issues there as there's no error handling involved (e.g. listing votes). Create vote could indeed be refactored, the [case statement ](https://github.com/liquidvotingio/api/blob/master/lib/liquid_voting_web/resolvers/voting.ex#L57-L77) feels like business logic to me, especially the call [to create the vote](https://github.com/liquidvotingio/api/blob/1c3ce5bf77dbf34328607e004c7c47ee9c049a60/lib/liquid_voting_web/resolvers/voting.ex#L95) and could be moved to another module as @oliverbarnes pointed out ☝️. 

But again, this isn't a crucial refactor as the separation of concerns is still fairly clear and the code is quite readable.
